Microsoft auth provider manages tokens for multiple resources

jamieklassen opened the pull request:
## Hey, I just made a Pull Request!

### Highlights

#### Backend (auth-backend plugin):
* Converted the tests for the `MicrosoftAuthProvider` to exercise a wider boundary, integrating with `OAuthAdapter`.
* When the `/frame/handler` endpoint is called with an authorization code for a non-Microsoft Graph scope, the user profile will not be fetched. Similarly no user profile or photo data will be fetched by the backend if the `/refresh` endpoint is called where the `scope` query parameter is missing the `User.Read` scope from Microsoft Graph.
#### Frontend (core-app-api package):
* when its `getAccessToken` method is called, `MicrosoftAuth` parses scopes to determine the audience of the token that azure will provision. Then it maintains a map of (audience, `OAuth2`) pairs so that each audience has its own notion of the current session.
* `MicrosoftAuth` always appends an `offline_access` scope to any token requests, so that tokens are always refreshable.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] ~Screenshots attached (for UI changes)~ n/a
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

jamieklassen commented:
It's been closed for staleness, but I believe this PR actually resolves #7040

jamieklassen commented:
@Rugvip politely pinging for review or reassignment on this one, which is depended upon by #16627 in order to solve #13685.

Rugvip commented on packages/core-app-api/src/apis/implementations/auth/microsoft/MicrosoftAuth.ts:
```diff
@@ -43,14 +60,108 @@ export default class MicrosoftAuth {
+    if (groups) {
+      const { resourceURI, scp } = groups;
+      const aud = resourceURI.replace(/^api:\/\//, '');
+      return { aud, scp };
```
Is `scp` short for scope? let's spell that out in that case :grin:, same thing with audience tbh, no need to keep things cryptic here. Same thing for the other occurrences.

jamieklassen replied:
I was obsessively simulating the exact claims that Azure uses for the sake of my tests -- all gone now

Rugvip commented on packages/core-app-api/src/apis/implementations/auth/microsoft/MicrosoftAuth.ts:
```diff
@@ -43,14 +60,108 @@ export default class MicrosoftAuth {
+      MicrosoftAuth.MicrosoftGraphID;
+    return {
+      aud: firstAudience,
+      scp: scopes
```
I don't see this return value used anywhere, should it be? Or should it be removed

jamieklassen replied:
It was being used only in tests, but now this function is much simplified and only extracts the resource from a set of scopes

Rugvip commented on plugins/auth-backend/src/providers/microsoft/provider.ts:
```diff
@@ -124,48 +143,57 @@ export class MicrosoftAuthProvider implements OAuthHandlers {
+          ),
+        }),
       }),
       refreshToken,
```
To double check this: we'll be issuing a lot of different tokens, but they're all stored in the same refresh token cookie. Is that fine? Does this refresh token refer to a session on the MS side that has the union of all scopes we've been requesting, even if they are for different resources?

Basically, if I first sign-in with MS auth, and then head over to the k8s cluster page and request access to those resources, I'd expect one auth popup for each of those steps. If I do the same thing again, I'd expect both of those authorizations to be part of the session already and neither of those two actions should trigger an auth popup again.

jamieklassen replied:
The answer to these question is a resounding 'Yes' -- Grants like authorization codes or refresh tokens are NOT subject to the same "only one resource" constraint that access tokens have. You can use a refresh token to get an access token for any consented scope from any resource, regardless of how you got that refresh token.

Hopefully the language in c15e0cedbe1981ccbe29bba83524d30b5357f74f makes this more clear:

> gets access token for other consented scopes besides those directly granted

Because this is really important -- it's the only reason any of this is viable!

Rugvip replied:
Excellent! 😁

Rugvip commented on plugins/auth-backend/src/providers/microsoft/fake.ts:
```diff
@@ -0,0 +1,126 @@
+
+type Claims = { aud: string; scp: string };
+
+export class FakeMicrosoftAPI {
```
This is just for tests right? You could place it in `plugins/auth-backend/src/providers/microsoft/__testUtils__/fake.ts` in that case.

jamieklassen replied:
b87d86c14b2878716db010c9b3249bda981ef3a0

Rugvip commented on packages/core-app-api/src/apis/implementations/auth/microsoft/MicrosoftAuth.ts:
```diff
@@ -43,14 +60,108 @@ export default class MicrosoftAuth {
+
+  static expectedClaims(scope: string): { aud: string; scp: string } {
+    const scopes = scope.split(' ').map(MicrosoftAuth.parseScope);
+    const firstAudience =
```
Rather than picking the first one and essentially leaving the rest broken, do you think it would make sense to be more strict and throw an error if there's a conflict in the requested scopes? I'm feeling that'll help users of this API figure out errors way earlier compared to having to figure out why their requested access token only has one of the two scopes requested. Frankly I feel this is something that it would make sense if the MS OAuth implementation would do ._.

That's if we can do this reliably and there's no valid case for passing conflicting scopes.

jamieklassen replied:
b76cbe84cc54983ffe95b67eaa308d928d649e23

Rugvip reviewed (changes requested):
Thank you, awesome stuff! 😁 🎉 

Really sweet to be able to avoid creeping this complexity to API consumers

jamieklassen commented:
@Rugvip I'm done for the week and want to push the changes I have, but I haven't finished addressing all your feedback. I'll ping again when it's truly ready

Rugvip reviewed (approved):
Awesome, thank you! 😁 👏 

Let's :shipit: 🚀
